File: js/4.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays = function(nums1, nums2) {
    let n = nums1.length, m = nums2.length, mn = n + m, inf = 0x7fffffff
    function kth(i, j, k) {
        let mid = k >> 1
        if(i >= n) return nums2[j + k - 1]
        if(j >= m) return nums1[i + k - 1]
        if(k == 1) return Math.min(nums1[i], nums2[j]);
        let r1 = i + mid - 1 >= n ? inf : nums1[i + mid - 1]
        let r2 = j + mid - 1 >= m ? inf : nums2[j + mid - 1]
        if(r1 < r2) return kth(i + mid, j, k - mid)
        return kth(i, j + mid, k - mid)
    }
    let mid = mn >> 1
    if(mn & 1) return kth(0, 0, mid + 1)
    else return (kth(0, 0, mid) + kth(0, 0, mid + 1)) / 2
};

File: go/4.go
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
    n := len(nums1)
    m := len(nums2)
    var i, j int
    k := (m + n - 1) >> 1
    for k > 0 {
        p := (k - 1) >> 1
        if j + p >= m || i + p < n && nums1[i + p] < nums2[j + p] {
            i += p + 1
        } else {
            j += p + 1
        }
        k -= p + 1
    }
    var s int
    if j >= m || i < n && nums1[i] < nums2[j] {
        s = nums1[i]
        i += 1
    } else {
        s = nums2[j]
        j += 1
    }
    if (n + m) % 2 == 1 {
        return float64(s)
    } else {
        if j >= m || i < n && nums1[i] < nums2[j] {
            return float64(s + nums1[i]) * 0.5
        } else {
            return float64(s + nums2[j]) * 0.5
        }
    }
}
